cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringServerHandler.java

```diff
@@ -0,0 +1,265 @@
+  private final TetheringStore store;
+  private final MessagingService messagingService;
+  private final MultiThreadMessagingContext messagingContext;
+  private final String topicPrefix;
+
+  @Inject
+  TetheringServerHandler(CConfiguration cConf, TetheringStore store, MessagingService messagingService) {
+    this.cConf = cConf;
+    this.store = store;
+    this.messagingService = messagingService;
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.topicPrefix = cConf.get(Constants.Tethering.TOPIC_PREFIX);
+  }
+
+  @Override
+  public void init(HandlerContext context) {
+    super.init(context);
+  }
+
+  /**
+   * Sends control commands to the client.
+   */
+  @GET
+  @Path("/tethering/controlchannels/{peer}")
+  public void connectControlChannel(FullHttpRequest request, HttpResponder responder, @PathParam("peer") String peer,
+                                    @QueryParam("messageId") String messageId)
+    throws IOException, NotImplementedException, PeerNotFoundException, ForbiddenException, BadRequestException {
+    checkTetheringServerEnabled();
+    store.updatePeerTimestamp(peer);
+    TetheringStatus tetheringStatus = store.getPeer(peer).getTetheringStatus();
+    if (tetheringStatus == TetheringStatus.PENDING) {
+      throw new PeerNotFoundException(String.format("Peer %s not found", peer));
+    } else if (tetheringStatus == TetheringStatus.REJECTED) {
+      responder.sendStatus(HttpResponseStatus.FORBIDDEN);
+      throw new ForbiddenException(String.format("Peer %s is not authorized", peer));
+    }
+
+    List<TetheringControlMessage> commands = new ArrayList<>();
+    MessageFetcher fetcher = messagingContext.getMessageFetcher();
+    TopicId topic = new TopicId(NamespaceId.SYSTEM.getNamespace(), topicPrefix + peer);
+    String lastMessageId = messageId;
+    try (CloseableIterator<Message> iterator =
+           fetcher.fetch(topic.getNamespace(), topic.getTopic(), 1, messageId)) {
+      while (iterator.hasNext()) {
+        Message message = iterator.next();
+        TetheringControlMessage controlMessage = GSON.fromJson(message.getPayloadAsString(StandardCharsets.UTF_8),
+                                                               TetheringControlMessage.class);
+        commands.add(controlMessage);
+        lastMessageId = message.getId();
+      }
+    } catch (TopicNotFoundException e) {
+      LOG.warn("Received control connection from peer {} that's not tethered", peer);
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(String.format("Invalid message id %s", messageId));
+    }
+
+    if (commands.isEmpty()) {
+      commands.add(new TetheringControlMessage(TetheringControlMessage.Type.KEEPALIVE));
+    }
+    TetheringControlResponse tetheringControlResponse = new TetheringControlResponse(lastMessageId, commands);
+    final InputStream in = new ByteArrayInputStream(GSON.toJson(tetheringControlResponse).getBytes());
```

Comment:
This change won't help memory consumption as it still buffer all `commands` in memory. You can revert to the old code to make it easier to read. Or if you want to switch to use BodyProducer, you need to use it to iterate the message iterator and produce content on demand.

Reply:
reverted to old code to keep it simple for now